llvm/tools/sycl-post-link/DeviceGlobals.h

```diff
@@ -0,0 +1,54 @@
+//===----- DeviceGlobals.h - SYCL Device Globals Pass ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// A transformation pass which converts symbolic device globals attributes
+// to integer id-based ones to later map to SPIRV device globals. The class
+// allso contains a number of static methods to extract corresponding
+// attributes of the device global variables and save them as a property set
+// for the runtime.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "llvm/ADT/MapVector.h"
+
+#include <cstdint>
+#include <vector>
+
+namespace llvm {
+
+class Module;
+class StringRef;
+
+// Represents a device global variable - at SYCL RT level device global
+// variables are being represented as a byte-array.
+struct DeviceGlobalProperty {
+  DeviceGlobalProperty(uint32_t Size, uint8_t DeviceImageScope)
+      : Size(Size), DeviceImageScope(DeviceImageScope) {}
+
+  // Encodes size of the underlying type T of the device global variable.
+  uint32_t Size;
+
+  // Either 1 (true) or 0 (false), telling whether the device global variable
+  // was declared with the device_image_scope property.
+  uint8_t DeviceImageScope;
+};
+
+using DeviceGlobalPropertyMapTy =
```

Comment:
Why do we need a `MapVector` here? Is there a need for a deterministic order of inserted elements here?

Reply:
The `PropertySetIO.h` header [defines a property set](https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/Support/PropertySetIO.h#L176) as a `MapVector`.

Reply:
> The `PropertySetIO.h` header [defines a property set](https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/Support/PropertySetIO.h#L176) as a `MapVector`. Yes, the deterministic order is required.

Ok, I see. I still don't fully understand why and where deterministic order is required, but if that's a `PropertySetIO` interface at the moment, then let's stick with it